api/types/authentication_authpreference_test.go

```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package types_test
+package types
```

Comment:
I'm not sure I follow the change - changing the test package should have no bearing on the initialization of the `types` package.

Does this actually fix the problems you are getting? `make test-api` works fine for me, I haven't been able to repro.

Reply:
I think this is the fact that we have symlink to the `teleport/api` package that teleport uses to vendor it, so there is something going on when resolving the packages as it re-evaluates the same package 2 times.

The `init` function is called from 2 different packages ((after adding `debug.PrintStack()`):

`"/teleport/api/types"`
`"/teleport/vendor/github.com/gravitational/teleport/api/types"`

I am trying to understand now why it does not repo on macs.

```
 go test -tags "   " github.com/gravitational/teleport/api/types

(calls init first time)

goroutine 1 [running, locked to thread]:
runtime/debug.Stack()
        /usr/local/go/src/runtime/debug/stack.go:24 +0x72
runtime/debug.PrintStack()
        /usr/local/go/src/runtime/debug/stack.go:16 +0x25
github.com/gravitational/teleport/api/types.init.0()
        /home/alexey/go/src/github.com/gravitational/teleport/api/types/types.pb.go:8085 +0x91


(calls init second time)

goroutine 1 [running, locked to thread]:
runtime/debug.Stack()
        /usr/local/go/src/runtime/debug/stack.go:24 +0x72
runtime/debug.PrintStack()
        /usr/local/go/src/runtime/debug/stack.go:16 +0x25
github.com/gravitational/teleport/vendor/github.com/gravitational/teleport/api/types.init.0()
        /home/alexey/go/src/github.com/gravitational/teleport/vendor/github.com/gravitational/teleport/api/types/types.pb.go:8085 +0x91
panic: proto: duplicate enum registered: types.PrivateKeyType
```

I am still investigating this one...

Reply:
This is puzzling. I tried reproing in Docker with both `golang:1.17` (Debian, if I'm not mistaken) and with the current `ubuntu` image, but no luck - `make test-api` work on both.

The change is harmless anyway, so leaving it approved in case you want to push it. Let us know when you have more info.

Reply:
I am not able to repro it Docker too, in docker the `api/` packages are always resolved against `/vendor` directory  when  running `api/.../*_tests.go` tests, it  hen follows the symlink -> `/go/src/github.com/gravitational/teleport/vendor/github.com/gravitational/teleport/api/types.init.0()`

On my machine, however,  when `api/./..._test.go` tests are run, the packages are resolved _locally_. No lookups are made to vendor.

```
github.com/gravitational/teleport/api/types.init.0()
       /go/src/github.com/gravitational/teleport/api/types/types.pb.go
```

But when I declare a test as a separate package e.i `types_test` then it starts resolving the `api` packages for _this_ test against `vendor` which causes multiple execution of `init` of api/types package resulting: 

```
panic: proto: duplicate enum registered: types.PrivateKeyType
```

@Joerger any ideas? :)

Reply:
Not Brian, but I may have a few ideas:

Which Go version are you running? I'm on latest / 1.17.4.

Maybe picking a specific vendor mode gives it a clue, if you haven't tried it already?  ( `cd api; go test -mod=vendor ./...`)?

Reply:
I'm not sure I understand the issue, `make test` and `make test-api` both work on my Ubuntu machine. Did you say you're running `go test ./api/types`? This fails for me with: `main module (github.com/gravitational/teleport) does not contain package github.com/gravitational/teleport/api/types` which is expected. 

`teleport` and `api` aren't run in the same test since they are separate modules, but if you are somehow managing to run them together (not sure how that would be possible), then you might get the double init issue you are describing.

Reply:
@Joerger I was running tests from root `$ make api-test` not from the api directory BUT I just noticed that I had
`GO111MODULE=off` set as my defaults so changing it to `auto` fixed the issue.  Thank you @codingllama!